File: reference-impl/typescript/test/agent.test.ts
import { describe, it, expect } from "vitest";
import { MockLanguageModelV3 } from "ai/test";
import { initialState } from "../src/domain";
import { Boundary, RecordingBus, fixedClock, sequentialIds, type Sink } from "../src/boundary";
import { runTurn } from "../src/agent";

// the verbose-but-required v6 usage block, factored out
const usage = {
  inputTokens: { total: 1, noCache: 1, cacheRead: undefined, cacheWrite: undefined },
  outputTokens: { total: 1, text: 1, reasoning: undefined },
};

describe("the runtime binding — the boundary is hooked onto onStepFinish", () => {
  it("runs a real generateText loop offline: the agent calls setPriority, the boundary folds it", async () => {
    // Script the model with NO network: step 1 calls setPriority; step 2 finishes.
    let call = 0;
    const model = new MockLanguageModelV3({
      doGenerate: async () => {
        call += 1;
        if (call === 1) {
          return {
            content: [
              {
                type: "tool-call",
                toolCallId: "c1",
                toolName: "setPriority",
                input: JSON.stringify({ ticket: "4118", level: "High" }),
              },
            ],
            finishReason: { unified: "tool-calls", raw: undefined },
            usage,
            warnings: [],
          };
        }
        return {
          content: [{ type: "text", text: "Priority set to High." }],
          finishReason: { unified: "stop", raw: undefined },
          usage,
          warnings: [],
        };
      },
    });

    const bus = new RecordingBus();
    const performed: string[] = [];
    const sink: Sink = { perform: (fx) => performed.push(fx.kind) };
    const boundary = new Boundary(
      fixedClock(7),
      sequentialIds("c"),
      bus,
      sink,
      initialState([{ id: "4118", body: "refund not received", priority: "Normal", status: { kind: "Open" } }]),
    );

    const out = await runTurn({ model, prompt: "ticket 4118 looks urgent", boundary });

    // the real loop ran two steps (tool call, then finish)
    expect(out.steps).toBe(2);
    // the boundary folded the agent's tool result → domain state updated
    expect(boundary.state.tickets.get("4118")!.priority).toBe("High");
    // committed as a signed Agent command, then the Log effect performed
    expect(bus.commands.map((c) => `${c.kind}:${c.by}`)).toContain("SetPriority:Agent");
    expect(performed).toContain("Log");
  });
});

File: reference-impl/kotlin/src/test/kotlin/adr/AgentTest.kt
package adr

import ai.torad.aisdk.providers.mockLanguageModelToolThenText
import ai.torad.aisdk.providers.mockToolInput
import kotlinx.coroutines.runBlocking
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class AgentTest {

    @Test
    fun `runs a real ToolLoopAgent loop offline - the agent calls setPriority, the boundary folds it`() =
        runBlocking {
            // Script the model with NO network: step 1 calls setPriority; step 2 finishes.
            val model = mockLanguageModelToolThenText(
                toolName = "setPriority",
                toolInput = mockToolInput("ticket" to "4118", "level" to "High"),
                finalText = "Priority set to High.",
            )

            val bus = RecordingBus()
            val performed = mutableListOf<Effect>()
            val sink = object : Sink {
                override fun perform(effect: Effect, mode: PerformMode) {
                    performed += effect
                }
            }
            val boundary = Boundary(
                Clock.fixed(7L),
                IdSource.sequential("c"),
                bus,
                sink,
                State.initial(
                    listOf(Ticket("4118", "refund not received", Priority.Normal, TicketStatus.Open)),
                ),
            )

            val agent = TriageAgent(model, boundary)
            val out = agent.runTurn("ticket 4118 looks urgent")

            // the real loop ran two steps (tool call, then finish)
            assertEquals(2, out.steps)
            // the boundary folded the agent's tool result → domain state updated
            assertEquals(Priority.High, boundary.state.tickets.getValue("4118").priority)
            // committed as a signed Agent command, then the Log effect performed
            assertTrue(bus.commands.any { it is Command.SetPriority && it.by == Actor.Agent })
            assertTrue(performed.any { it is Effect.Log })
        }
}
